text/0060-bazel.md

```diff
@@ -0,0 +1,137 @@
+
+Instead of running `bash build.sh` or `npm run build` on a subpackage, an engineer would simply run `bazel build <path-to-resource>` or
+`bazel test <path-to-resource`. Running a specific Bazel rule (for instance orchestrating a release) is also straightforward, with
+`bazel run <path-to-resource>.<action>`.
+
+# Design Summary
+
+In order to setup Bazel, the Bazel-specific rules and resources need to be added to the repository. This means adding `BUILD.bazel` files
+to each subpackage, and then building custom rules around some of the internals of the project (e.g. CFN to TS).
+
+# Detailed Design
+
+The structure of adding Bazel to a project is quite simple. Represented below is a tree of the files
+needed for a standard project
+
+```
+BAZEL.build
+package.json
+|packages|
+  BUILD.bazel
+  |apigateway|
+    BUILD.bazel
+    package.json
+    apigateway.ts
+    index.ts
+WORKSPACE <-- Bazel project definition, only needed once
+yarn.lock
+```
+
+The `BUILD` files themselves are quite simple, since Bazel has rules built-in for compiling a TypeScript library and for
+running JavaScript tests. We would need to write custom rules in the 
+[Skylark language](https://docs.bazel.build/versions/master/skylark/language.html) to perform custom operations like
+building TypeScript definitions from CloudFormation definition files. This is relatively straightforward since the tooling
+for this exists in TypeScript already, and Skylark is a subset of Python.
+
+Example `BUILD.bazel` file, assuming `cdk_library` is a Bazel rule that calls `ts_library` with another rule that performs 
+TS transformation:
+```
+load("//tools:defaults.bzl", "cdk_library", "pkg_npm")
+
+package(default_visibility = ["//visibility:public"])
+
+cdk_library(
+    name = "aws-amazonmq",
+    scope = ["AWS::AmazonMQ"],
+    outputs = ["lib/amazonmq.generated.ts"],
+    srcs = glob(["lib/**/*.ts"]),
+    deps = [
+        "//packages/@aws-cdk/cdk",
+    ],
+)
+
+pkg_npm(
+    name = "npm_package",
+    srcs = [
+      "package.json",
+    ],
+    tags = [
+        "release-with-framework",
+    ],
+    deps = [
+        ":aws-amazonmq",
+    ],
+)
+```
+
+The `pkg_npm` rule is also built-in to Bazel, and bundles a package based on given sources, along with other needed files like a package.json.
+Note that the `cdk_library` rule is only needed for packages that have a CFN to TS transformation. If that's not needed, it's simply a `ts_library` rule.
+
+# Drawbacks
+
+There are two main drawbacks to migrating to Bazel:
+
+1. Lack of familiarity with Bazel amongst the team
+1. Relative newness of Bazel in the JavaScript/TypeScript ecosystem
+
+For the first point, the current system has the advantage of being a glorified monorepo wrapper around NPM. Bazel,
+on the other hand, has a completely different semantic structure. While using it is straightforward for the most part,
+understanding how it works and more importantly how to extend it (by writing custom rules, for example) is not.
```

Comment:
I would like this statement to be more substantiated. I *think* we're doing stuff in a way that is very much "how you do it in the node ecosystem"; but I'm also no expert on the subject (I mean, AWS CDK is the first ever `node` project I work on) and am willing to believe I may be misled by my own perceptions.

Reply:
`Lerna`, while certainly a capable tool, is also a stagnant one. In essence, it's a glorified symlink manager for grouped dependencies in one repository. At one point in time, this was excellent and all that was needed for Node development. However, the size and scope of Node projects in general has increased in complexity, while Lerna's capabilities have not. To a certain degree, Yarn workspaces are trying to help with the innovative process, but it has a long way to go.

I'm not saying that Bazel is "the way to do things" in the JavaScript world (not yet, anyways), but it certainly has its advantages for a sprawling monorepo like the CDK. This would be very different if the CDK were a single module, but it's not. Bazel was built to solve the issue of tons of modules in various states across developers. Lerna was build to solve the issue of a developer with a couple related libraries housed in one repository.